From d4c9d409e1b98ff25edf098d46329e945df1398b Mon Sep 17 00:00:00 2001
From: Phil Weaver <pweaver@google.com>
Date: Wed, 12 Jul 2017 14:07:07 -0700
Subject: [PATCH 6/8] Back-port ag/2491664

Bug: 62196835
Test: Verify overlays disappear on a11y capabilities
dialog.

Change-Id: Ic675012dd9faa8e53d1d4b126b3ba68fecdab992
(cherry picked from commit e76e053595660ed2a75adb33ee124a7bfbed164b)
CVE-2017-0752
---
 AndroidManifest.xml                                |  1 +
 ...ggleAccessibilityServicePreferenceFragment.java | 37 ++++++++++++++--------
 2 files changed, 24 insertions(+), 14 deletions(-)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index f746f2b..7a11311 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -67,6 +67,7 @@
     <uses-permission android:name="android.permission.ACCESS_NOTIFICATIONS" />
     <uses-permission android:name="android.permission.REBOOT" />
     <uses-permission android:name="android.permission.MANAGE_DEVICE_ADMINS" />
+    <uses-permission android:name="android.permission.HIDE_NON_SYSTEM_OVERLAY_WINDOWS" />
 
     <application android:label="@string/settings_label"
             android:icon="@mipmap/ic_launcher_settings"
diff --git a/src/com/android/settings/accessibility/ToggleAccessibilityServicePreferenceFragment.java b/src/com/android/settings/accessibility/ToggleAccessibilityServicePreferenceFragment.java
index 0c568f0..35ec790 100644
--- a/src/com/android/settings/accessibility/ToggleAccessibilityServicePreferenceFragment.java
+++ b/src/com/android/settings/accessibility/ToggleAccessibilityServicePreferenceFragment.java
@@ -31,6 +31,8 @@ import android.provider.Settings;
 import android.text.TextUtils;
 import android.view.LayoutInflater;
 import android.view.View;
+import android.view.Window;
+import android.view.WindowManager;
 import android.view.accessibility.AccessibilityManager;
 import android.widget.ImageView;
 import android.widget.LinearLayout;
@@ -44,6 +46,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
+import static android.view.WindowManager.LayoutParams.PRIVATE_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS;
+
 public class ToggleAccessibilityServicePreferenceFragment
         extends ToggleFeaturePreferenceFragment implements DialogInterface.OnClickListener {
 
@@ -154,21 +158,26 @@ public class ToggleAccessibilityServicePreferenceFragment
     public Dialog onCreateDialog(int dialogId) {
         switch (dialogId) {
             case DIALOG_ID_ENABLE_WARNING: {
-            mShownDialogId = DIALOG_ID_ENABLE_WARNING;
-            AccessibilityServiceInfo info = getAccessibilityServiceInfo();
-            if (info == null) {
-                return null;
+                mShownDialogId = DIALOG_ID_ENABLE_WARNING;
+                AccessibilityServiceInfo info = getAccessibilityServiceInfo();
+                if (info == null) {
+                    return null;
+                }
+                AlertDialog ad = new AlertDialog.Builder(getActivity())
+                        .setTitle(getString(R.string.enable_service_title,
+                                info.getResolveInfo().loadLabel(getPackageManager())))
+                        .setIconAttribute(android.R.attr.alertDialogIcon)
+                        .setView(createEnableDialogContentView(info))
+                        .setCancelable(true)
+                        .setPositiveButton(android.R.string.ok, this)
+                        .setNegativeButton(android.R.string.cancel, this)
+                        .create();
+                Window window = ad.getWindow();
+                WindowManager.LayoutParams params = window.getAttributes();
+                params.privateFlags |= PRIVATE_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS;
+                window.setAttributes(params);
+                return ad;
             }
-            return new AlertDialog.Builder(getActivity())
-            .setTitle(getString(R.string.enable_service_title,
-                    info.getResolveInfo().loadLabel(getPackageManager())))
-                    .setIconAttribute(android.R.attr.alertDialogIcon)
-                    .setView(createEnableDialogContentView(info))
-                    .setCancelable(true)
-                    .setPositiveButton(android.R.string.ok, this)
-                    .setNegativeButton(android.R.string.cancel, this)
-                    .create();
-        }
             case DIALOG_ID_DISABLE_WARNING: {
             mShownDialogId = DIALOG_ID_DISABLE_WARNING;
             AccessibilityServiceInfo info = getAccessibilityServiceInfo();
-- 
2.7.4

